README.md
# hypoDDpy

This a collection of tools to run
[HypoDD](http://www.ldeo.columbia.edu/~felixw/hypoDD.html) by Felix Waldhauser,
updated by Wayne Crawford.

It takes event files in the QuakeML format, station data in the SEED format and
waveform data in any format ObsPy can read and does all the rest.

The output is one QuakeML file with the relocated events having one additional
Origin node. The events that could not be relocated will not be changed.

![Flowchart 1](https://raw.github.com/krischer/hypoDDpy/master/img/flowchart.png)

### Did It Help You in Your Research?

If yes, consider citing this software! Just click on this button here to get all the required information:

[![DOI](https://zenodo.org/badge/doi/10.5281/zenodo.18907.svg)](http://dx.doi.org/10.5281/zenodo.18907)


### Requirements

* Python 3.6+
* NumPy
* matplotlib
* progressbar
* cartopy
* [ObsPy](http://obspy.org) (Tested on 1.2.0)

### Installation
hypoDDpy currently works with HypoDD 2.1b which you will have to acquire
separately from Felix Waldhauser.

Put the archive here:
```
hypoddpy/src/HYPODD_2.1b.tar.gz
```
The src directory will likely not exist.

Then run *either of the* following two commands, depending on which Python module installer you prefer:
```
pip install -v -e .
python setup.py develop
```

The in-place install is a good idea because there is a chance that you will
have to adjust the source code.


### Running it

It is steered via a Python script that you will have to create. It should be
rather self-explanatory.

After you created it, simply run it to perform the relocation.

```python
import glob
from hypoddpy import HypoDDRelocator


# Init the relocator with the working directory and some necessary
# configuration values.
#
# The working dir is where all the working files and some output files will be
# stored.
# All the other attributes are related to the cross correlation and should be
# self-explanatory.
relocator = HypoDDRelocator(working_dir="relocator_working_dir",
    cc_time_before=0.05,
    cc_time_after=0.2,
    cc_maxlag=0.1,
    cc_filter_min_freq=1.0,
    cc_filter_max_freq=20.0,
    cc_p_phase_weighting={"Z": 1.0},
    cc_s_phase_weighting={"Z": 1.0, "E": 1.0, "N": 1.0},
    cc_min_allowed_cross_corr_coeff=0.4)

# Add the necessary files. Call a function multiple times if necessary.
relocator.add_event_files(glob.glob("events/*.xml"))
relocator.add_waveform_files(glob.glob("waveform/*.mseed"))
relocator.add_station_files(glob.glob("station/*.xml"))

# Setup the velocity model. This is just a constant velocity model.
relocator.setup_velocity_model(
    model_type="layered_p_velocity_with_constant_vp_vs_ratio",
    layer_tops=[(-10000, 5.8)],
    vp_vs_ratio=1.73)

# Start the relocation with the desired output file.
relocator.start_relocation(output_event_file="relocated_events.xml")

# Plot events with a slightly better presentation than the default plots
# Have to use "replace_scatter_with_plot" until cartopy 0.18.1 comes out
relocator.plot_events(coastlines='10m', replace_scatter_with_plot=True)
```

You can also plot the results after the fact, using the output event file:
```python
from hypoddpy import HypoDDPlotter
from obspy.core.event import read_events

catalog_file = 'relocated_events.xml'
map_extent=[45.15, 45.75, -13, -12.6]
depth_extent=[0, 50]

print(f'Reading catalog file {catalog_file}...', end='', flush=True)
cat = read_events(catalog_file, 'QUAKEML')
print(f'read {len(cat)} events')
obj = HypoDDPlotter(cat, map_extent, depth_extent, coastlines='10m',
                    replace_scatter_with_plot=True)
obj.plot_events(file_base='main_swarm_EW',
                polygon=[[45.3, -12.78],
                         [45.3, -12.82],
                         [45.5, -12.82],
                         [45.5, -12.78]])
obj.plot_events(file_base='main_swarm_NS',
                polygon=[[45.39, -12.65],
                         [45.39, -12.95],
                         [45.42, -12.95],
                         [45.42, -12.65]])
obj.plot_events(file_base='secondary_swarm',
                polygon=[[45.50, -12.70],
                         [45.65, -12.80],
                         [45.65, -12.90],
                         [45.50, -12.80]])
```


CHANGELOG.md
v0.0.2:
======================

* Works on python 3.6+
* Added StationXML file read to HypoDDRelocator._parse_station_files()
* If {net}.{sta} > 7 characters (hypoDD limitation), only puts {sta} in
  input files
* fixed distance dimensions bug in _write_ph2dt_inp_file() (maxsep was 
  calculated using depth differences in meters instead of km)
- Added check for > 30 model layers to HypoDDRelocator.setup_velocity_model()
- Added shift_stations parameter to HypDDRelocator class, to overcome
  HypoDD limitation that any stations with negative elevation are shifted to
  zero elevation.  Shifts stations and model so that minimum station elevation
  is zero, then shifts event depth back when writing to QuakeML file (SHIFTING
  BACK ONLY WORKS IF THE station.json file didn't already exist, so if you just
  modify hypoDD.inc and rerun the depths will not be shifted: see TODO.md).
- Added  HypoDDRelocator.modify_compile_parameters(): allows the user to
  specify parameters used to compile hypoDD
- Added class HypoDDPlotter, which makes more geometrically accurate plots
- Added method hypodd_relocator.plot_events(), which uses HypoDDPlotter
  to plot events after they are relocated

hypoddpy/hypodd_plotter.py
#!/usr/bin/env python3
"""
Creates some plots of the relocated event Catalog.

Colors (defined using ColorConverter and get_cmap) distinguish clusters.
    grey: non-relocated
    red: relocated with cluster id 1
    green: relocated with cluster id 2
    blue: relocated with cluster id 3
    yellow: relocated with cluster id 4
    orange: relocated with cluster id 5
    cyan: relocated with cluster id 6
    magenta: relocated with cluster id 7
    brown: relocated with cluster id 8
    lime: relocated with cluster id >8 or undetermined cluster_id
"""
import cartopy.crs as ccrs
import matplotlib.pylab as plt
from matplotlib.cm import get_cmap
from matplotlib.colors import ColorConverter
from obspy.core.event import read_events

from .plotting.shape_selection import Polygon, Circle
from .plotting.map_cross_section import MapCrossSection


class Positions():
    """
    Simple class for EQ Positions
    """
    def __init__(self):
        """
        Set up parameters
        """
        self.latitudes = []
        self.longitudes = []
        self.depths = []

    def validate(self):
        """Returns true if all lists are same length"""
        return (len(self.latitudes) == len(self.longitudes)
                and len(self.latitudes) == len(self.depths))


class HypoDDPlotter():
    """
    Plot original and relocated events from HypoDDpy
    """

    def __init__(self, cat, map_extent=None, depth_extent=None,
                 coastlines=None, replace_scatter_with_plot=False,
                 quiet=False):
        """
        Set up parameters

        :param cat: obspy catalog.  Original events should be in origins[0],
            relocated events in origins[-1]
        :param map_extent: limits of map to plot [W, E, S, N].  If None, uses
            limits of event lon,lats plus a small buffer
        :param depth_extent: depth limits to plot [mindepth, maxdepth].  If
            None, uses limits of event depths, plus a small buffer
        :param coastlines: The name of a NaturalEarth ('10m', '50m', '110m')
            or GSHSS ('crude', 'low', 'intermediate', 'high', 'full') coastlines
            dataset to use.
        :param replace_scatter_with_plot: Use matplotlib.pyplot.plot() instead
            of matplotlib.pyplot.scatter() (temporary fix to avoid cartopy
            0.18 bug that should be fixed in cartopy 0.18.1)
        :param quiet: Do not use print statements
        """
        self.cat = cat
        self.map_extent = map_extent
        self.depth_extent = depth_extent
        self.coastlines = None
        self.file_base = 'hypoDDpy'
        self.original_filename = ""
        self.relocated_filename = ""
        self.replace_scatter_with_plot=replace_scatter_with_plot
        self.quiet = quiet

    def plot_events(self, file_base='hypoDDpy', circle=None, polygon=None,
                    coastlines=None):
        """
        Plot original and relocated events
        :param file_base: start of output filename
        :param polygon: list of [lon, lat]s defining a polygon to highlight
        :param circle: [lon, lat, radius_km] of a circle to highlight (will
            be ignored if polygon defined)
        :param coastlines: as in constructor
        """
        if coastlines:
            self.coastlines = coastlines
        if file_base:
            self.file_base = file_base
        shape=None
        if circle:
            shape=Circle(circle)
        if polygon:
            shape=Polygon(polygon)
        self._plot_original_relocated(shape=shape)
        return self.original_filename, self.relocated_filename

    def _plot_original_relocated(self, shape=None, marker_size=1):
        """
        Plot original and relocated events

        :param shape: a shape to plot and highlight events within
        :param marker_size:
        """
        self.original_filename = f"{self.file_base}_original.pdf"
        self.relocated_filename = f"{self.file_base}_relocated.pdf"

        original_pos, relocated_pos = self._extract_positions()
        self._set_plot_params(shape)
        self.marker_size = marker_size

        # Plot the original event locations.
        map_extent, depth_extent = self._plot_events(
            original_pos,
            shape=shape,
            map_extent=self.map_extent,
            depth_extent=self.depth_extent)
        plt.savefig(self.original_filename)
        if not self.quiet:
            print("Output figure: %s" % self.original_filename)

        # Plot the relocated event locations with same extents as original
        self._plot_events(
            relocated_pos,
            shape=shape,
            map_extent=map_extent,
            depth_extent=depth_extent)
        plt.savefig(self.relocated_filename)
        if not self.quiet:
            print("Output figure: %s" % self.relocated_filename)

    def _extract_positions(self):
        """
        Extract original and relocated Positions for each event
        """
        original_pos = Positions()
        relocated_pos = Positions()

        for event in self.cat:
            # The first origin is always the original
            original_pos.latitudes.append(event.origins[0].latitude)
            original_pos.longitudes.append(event.origins[0].longitude)
            original_pos.depths.append(event.origins[0].depth / 1000.0)
            
            # The last origin could  be a relocated one or an original one.
            relocated_pos.latitudes.append(event.origins[-1].latitude)
            relocated_pos.longitudes.append(event.origins[-1].longitude)
            relocated_pos.depths.append(event.origins[-1].depth / 1000.0)
        return original_pos, relocated_pos
                    
    def _set_plot_params(self, shape):
        """
        Set plot parameters for each event
        """
        color_invalid = ColorConverter().to_rgba("grey")
        cmap = get_cmap("Paired", 12)

        self.colors = []
        self.magnitudes = []
        self.inshape = []

        for event in self.cat:
            self.magnitudes.append(event.magnitudes[0].mag)
            # Use color to Code the different events. Colorcode by event
            # cluster or indicate if an event did not get relocated.
            if event.origins[-1].method_id is None or \
               "HYPODD" not in str(event.origins[-1].method_id).upper():
                self.colors.append(color_invalid)
            # Otherwise get the cluster id, stored in the comments.
            else:
                for comment in event.origins[-1].comments:
                    comment = comment.text
                    if comment and "HypoDD cluster id" in comment:
                        cluster_id = int(comment.split(":")[-1])
                        break
                else:
                    cluster_id = 0
                self.colors.append(cmap(int(cluster_id)))
            self.inshape.append(True)
            if shape:
                if not shape.contains(event.origins[-1].longitude,
                                      event.origins[-1].latitude):
                    self.inshape[-1] = False
                    
    def _plot_events(self, positions, shape,
                     map_extent=None, depth_extent=None):
        """
        Plot event locations as map view and 2 cross-sections

        Uses cartopy
        :param positions:  Positions object
        :param shape:
        :param map_extent:
        :param depth_extent:
        """
        plt.clf()
        mcs = MapCrossSection(map_extent=map_extent,
                              depth_extent=depth_extent,
                              latitudes=positions.latitudes,
                              longitudes=positions.longitudes,
                              depths=positions.depths,
                              fill_figure=False,
                              coastlines=self.coastlines)

        self._plot_2D(mcs.ax_map, positions.longitudes, positions.latitudes,
                      transform=ccrs.Geodetic())
        if shape:
            shape.plot(mcs.ax_map, linestyle='--', linewidth=0.5,
                       transform=ccrs.Geodetic())

        # Latitude cross-section
        self._plot_2D(mcs.ax_zy, positions.depths, positions.latitudes)
        mcs.ax_zy.set_xlabel('Depth (km)')
        mcs.ax_zy.set_ylabel('Latitude')

        # Longitude cross-section
        self._plot_2D(mcs.ax_xz, positions.longitudes, positions.depths)
        mcs.ax_xz.set_xlabel('Longitude')
        mcs.ax_xz.set_ylabel('Depth (km)')

        return mcs.map_extent, mcs.depth_extent

    def _plot_2D(self, ax, xs, ys, ref_mag=None, transform=None):
        """
        Plot a 2D view (map or cross-section)

        :param ax: axes object
        :param xs: x-axis values
        :param ys: y-axis values (same length as xs)
        :param ref_mag: magnitude at which the marker size = marker_size.  If
            None, use marker_size everywhere.
        :param transform: transform to use when plotting (None for
            cross-sections, ccrs.Geodetic() for map view)
        """
        if ref_mag:
            sizes = [self.marker_size*3**(x-ref_mag) for i, x in
                     zip(self.inshape, self.magnitudes) if i is True]
        else:
            sizes = self.marker_size**2
        if transform is None:
            transform = ax.transData
        if self.replace_scatter_with_plot:
            ax.plot(xs, ys, color='#e0e0e0', marker='.', markersize=0.1,
                    linestyle='', transform=transform)
            ax.plot([x for i, x in zip(self.inshape, xs) if i is True],
                    [y for i, y in zip(self.inshape, ys) if i is True],
                    color='k',
                    marker='.',
                    markersize=0.1,
                    linestyle='',
                    transform=transform)
        else:
            ax.scatter(xs, ys, s=sizes, c='#e0e0e0', transform=transform)
            ax.scatter([x for i, x in zip(self.inshape, xs) if i is True],
                       [x for i, x in zip(self.inshape, ys) if i is True],
                       s=sizes,
                       c=[x for i, x in zip(self.inshape, self.colors)
                          if i is True],
                       marker='.',
                       transform=transform)


if __name__ == "__main__":
    catalog_file = 'relocated_events.xml'
    print(f'Reading catalog file {catalog_file}...', end='', flush=True)
    cat = read_events(catalog_file, 'QUAKEML')
    print(f'read {len(cat)} events')
    obj = HypoDDPlotter(cat, map_extent=[45.15, 45.75, -13, -12.6],
                             depth_extent=[0, 50])
    obj.plot(file_base='All', coastlines='10m')
    obj.plot(file_base='main_swarm_EW', coastlines='10m',
             polygon=[[45.3, -12.78],
                      [45.3, -12.82],
                      [45.5, -12.82],
                      [45.5, -12.78]])
    obj.plot(file_base='main_swarm_NS', coastlines='10m',
             polygon=[[45.39, -12.65],
                      [45.39, -12.95],
                      [45.42, -12.95],
                      [45.42, -12.65]])
    obj.plot(file_base='secondary_swarm', coastlines='10m',
             polygon=[[45.50, -12.70],
                      [45.65, -12.80],
                      [45.65, -12.90],
                      [45.50, -12.80]])
